File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Action/Common/DoFuncAction.ts
import Async from "../../../../../Core/Async";
import AbstractGameAction from "../AbstractGameAction";

export class DoFuncAction extends AbstractGameAction{
    public constructor(private Func:Function,private Interval:number = 0) {
        super();
    }

    async Update(): Promise<boolean> {
        if(this.Duration == 0){
            this.Func()
        }
        if(this.Interval == 0)
            return true;
        await Async.Delay(this.Interval);
        return true;
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Action/Utility/DoFuncAction.cs
using System;
using Cysharp.Threading.Tasks;
using ModTheSpire.Event;

namespace ModTheSpire.Effect.Other
{
    public class DoFuncAction : IGameAction
    {
        public float Duration { get; set; }
        public ActionType Type { get; } = ActionType.WAIT;
        private int Interval { get; }
        private Action Action { get; }
        public void Init()
        {
        }

        public DoFuncAction(Action action,int duration = 0)
        {
            Action = action;
            Interval = duration;
        }

        public async UniTask<bool> Update()
        {
            Action();
            await UniTask.Delay(Interval);
            return true;
        }
    }
}